### PR TITLE
expose BasicDeliverEventArgs in message context

### PR DESCRIPTION
**Pull request by SimonCropp**
proposal for https://github.com/Particular/NServiceBus.RabbitMQ/issues/103

### CODE REVIEW

**Comment by SimonCropp** [comment]
@Particular/rabbitmq-transport-maintainers thoughts?

**Comment by adamralph** [comment]
Thanks @SimonCropp, I'll take a look ASAP.

File: src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs [hunk 1]
```diff
@@ -214,6 +214,7 @@ void Connection_ConnectionShutdown(object sender, ShutdownEventArgs e)
                 var errorHandled = false;
                 var numberOfDeliveryAttempts = 0;
 
+                contextBag.Set(message);
```
Comment (bording) [inline src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs]: @SimonCropp FYI, for this to work the way you're intending, you'll have to also undo the optimization currently in place around the creation of the `ContextBag`. Currently, because we don't use the bag for anything, there is a single static instance that is reused for every message, saving having to allocate one per message.

Comment (SimonCropp) [inline src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs]: > you'll have to also undo the optimization currently in place around the creation of the ContextBag

how does that work when other things place things in the context? eg https://docs.particular.net/nservicebus/sql-persistence/#sqlstoragesession

Comment (bording) [inline src/NServiceBus.RabbitMQ/Receiving/MessagePump.cs]: The ContextBag created by the transport gets merged into the ContextBag managed by Core.

File: src/NServiceBus.RabbitMQ.Tests/APIApprovals.cs [hunk 2]
```diff
@@ -13,7 +14,9 @@ public class APIApprovals
     [MethodImpl(MethodImplOptions.NoInlining)]
     public void Approve()
     {
-        var publicApi = Filter(ApiGenerator.GeneratePublicApi(typeof(RabbitMQTransport).Assembly));
+        var combine = Path.Combine(TestContext.CurrentContext.TestDirectory, "NServiceBus.Transports.RabbitMQ.dll");
+        var assembly = Assembly.LoadFile(combine);
+        var publicApi = Filter(ApiGenerator.GeneratePublicApi(assembly));
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ.Tests/APIApprovals.cs]: What is the purpose of this change?

File: src/NServiceBus.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs [hunk 3]
```diff
@@ -128,5 +128,13 @@ public static TransportExtensions<RabbitMQTransport> SetClientCertificates(this
             transportExtensions.GetSettings().Set(SettingsKeys.ClientCertificates, clientCertificates);
             return transportExtensions;
         }
+
+        /// <summary>
+        /// Gets the <see cref="BasicDeliverEventArgs"/> for the current message being processed.
+        /// </summary>
+        public static BasicDeliverEventArgs GetBasicDeliverEventArgs(this IMessageHandlerContext context)
```
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs]: I'd prefer to put this in a `MessageHandlerContextExtensions` class.

Comment (SimonCropp) [inline src/NServiceBus.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs]: i figured it want worth leaking a new type for one method when they will never see that class anyway

File: src/NServiceBus.RabbitMQ.Tests/APIApprovals.cs (hunk 2 above)
Comment (SimonCropp) [inline src/NServiceBus.RabbitMQ.Tests/APIApprovals.cs]: because given your test runner typeof(RabbitMQTransport).Assembly might resolve to somewhere where dependency dlls do not exist. it was failing for me

Comment (adamralph) [inline src/NServiceBus.RabbitMQ.Tests/APIApprovals.cs]: 👍 good to know.

File: src/NServiceBus.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs (hunk 3 above)
Comment (adamralph) [inline src/NServiceBus.RabbitMQ/Configuration/RabbitMQTransportSettingsExtensions.cs]: OK. It's a minor point anyway. We can always change later.

**Comment by adamralph** [comment]
As [mentioned in the issue](https://github.com/Particular/NServiceBus.RabbitMQ/issues/103#issuecomment-281376218), @Particular/rabbitmq-transport-maintainers feel that the best way to do this is to provide a config extension which enables the propagation of `BasicDeliverEventArgs` to the handler. By default, the object will not be propagated and the current optimisation will remain, using the singleton `ContextBag`. When the the config extension is enabled, the optimisation will bypassed, and a `ContextBag` will be created for the scope of the message, containing the `BasicDeliverEventArgs`. The extension method should log a warning when the `BasicDeliverEventArgs` does not exist in the `ContextBag`, indicating that this config switch should be used to enable the propagation.

**Comment by SimonCropp** [comment]
Ok. Sounds like this is not required then

**Comment by adamralph** [comment]
@SimonCropp we could have built on top of your changes. They were a step towards the solution. Sorry for not making that clear.

**Comment by adamralph** [comment]
I stumbled upon the commits in my local clone 😄 - https://github.com/Particular/NServiceBus.RabbitMQ/pull/353
